File: Solutions/PE001_multiples_of_3_and_5/multiples_of_3_and_5.py
#!/bin/python3

# https://www.hackerrank.com/contests/projecteuler/challenges/euler001

# formula for sum of numbers: https://brilliant.org/wiki/sum-of-n-n2-or-n3/


def sum_1_to(n):
    return n*(n+1)//2


def multiples_of_3_and_5(n):
    """
    return sum of multiples of 3 and 5 below n
    """
    number_of_multiples_of_3 = (n-1)//3
    number_of_multiples_of_5 = (n-1)//5
    number_of_multiples_of_15 = (n-1)//15
    sum_of_multiples_of_3 = 3*sum_1_to(number_of_multiples_of_3)
    sum_of_multiples_of_5 = 5*sum_1_to(number_of_multiples_of_5)
    sum_of_multiples_of_15 = 15*sum_1_to(number_of_multiples_of_15)
    return sum_of_multiples_of_3 + sum_of_multiples_of_5 - sum_of_multiples_of_15


if __name__ == '__main__':
    # print(multiples_of_3_and_5(1000))
    pass

File: Solutions/PE001_multiples_of_3_and_5/cpp/MultiplesOf3And5.cpp
#include <iostream>

//https://www.hackerrank.com/contests/projecteuler/challenges/euler001

using namespace std;

long int SumOneTo(long int n){
    return n*(n+1)/2;
}

long int MultiplesOf3And5(long int n){
    long int NumberOfMultiplesOf3, NumberOfMultiplesOf5, NumberOfMultiplesOf15;
    long int SumOfMultiplesOf3, SumOfMultiplesOf5, SumOfMultiplesOf15;
    NumberOfMultiplesOf3 = (n-1)/3;
    NumberOfMultiplesOf5 = (n-1)/5;
    NumberOfMultiplesOf15 = (n-1)/15;
    SumOfMultiplesOf3 = 3*SumOneTo(NumberOfMultiplesOf3);
    SumOfMultiplesOf5 = 5*SumOneTo(NumberOfMultiplesOf5);
    SumOfMultiplesOf15 = 15*SumOneTo(NumberOfMultiplesOf15);
    return SumOfMultiplesOf3 + SumOfMultiplesOf5 - SumOfMultiplesOf15;
}

int main()
{
    int t;

    cin >> t;
    for(int a0 = 0; a0 < t; a0++){
        int n;
        cin >> n;
        cout << MultiplesOf3And5(n) << endl;
    }
    return 0;
}
